### ISSUE TITLE
['config']['environment'] = false should be default

### ISSUE THREAD

**Issue by bdangit** [opscode-cookbooks/chef-client#310]
Had a hard time figuring out why my nodes were modifying its environment when I explicitly move them to a new chef environment via Knife.  This change that was introduced via **Cookbook doesn't seem to pick up bootstrap'd "environment" #262** has caused a lot of headaches.

Would love to understand the rationale of why the default action is to grab whatever chef environment they exist in currently on the first converge of using `recipe[chef-client::config]`.  As it appears that once the `environment` parameter is defined inside `/etc/chef/client.rb` it will override anyone trying to migrate the node into a new environment via `knife exec` node transforms.  *This might be a good thing as I write this issue.*

**Comment by jf647** [comment]
We never bake the environment into our client.rb files, as our nodes come up in a bootstrap environment and then get specialized after chef-vault encrypted data bags have been refreshed for them.

As a mechanism of enforcing policy, this is fairly weak, especially given that anyone with write access to the node (which they would need to change the environment) could just as easily set a ['config']['environment'] to false at the node level to get around it.

For the people who have to give access to possibly untrusted people, there are other solutions available (Chef Analytics for example can trigger on node edits that match a particular rule).

+1 on not making the environment sticky based on what was used when the node was bootstrapped.

**Comment by jtimberman** [comment]
I can't speak to the motivation for the change in issue #262, or PR #267. However, it appears that you can override that behavior by setting the attribute to `false` (literal keyword, not a string `"false"`). For example in a cookbook or [Policyfile](https://github.com/chef/chef-dk/blob/master/POLICYFILE_README.md#define-attributes):

```ruby
default['chef_client']['config']['environment'] = false
```

In a role:

```json
{
  "default_attributes": {
    "chef_client": {
      "config": {
        "environment": false
      }
    }
  }
}
```

**Comment by juliandunn** [comment]
Sure, I can buy the arguments to roll it back; if we consider that the Chef server is supposed to be the canonical source of "policy" for the nodes, then it makes sense. Heads up @jfbibeau

**Comment by jfbibeau** [comment]
Aww man, I just upgraded to chef-client 4.3.0 with the fix in #267 and I was all happy :)

Alright, I think I can live with it if you roll it back. I was under the impression that without the environment in client.rb, the periodic chef-daemon run would not pick up the environment (ie: assume it was default). I just did a quick bit of testing and it does seem that as long as you setup the node with an environment it doesn't need to actually be populated in client.rb.

Thanks for the heads up.
